### PR TITLE
[TKET-1567] Add aas in rv3

### PR DESCRIPTION
**Pull request by cqc-melf**

### CODE REVIEW

**Comment by cqc-melf** [comment]
@sjdilkes if you want to and you have the time, you can take a first look.

File: tket/src/Mapping/AASRoute.cpp [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+    if (mapping_frontier->circuit_.get_OpType_from_Vertex(v) ==
+        OpType::PhasePolyBox) {
+      TKET_ASSERT(mapping_frontier->circuit_.is_quantum_node(v));
+      Op_ptr op_ptr_ppb = mapping_frontier->circuit_.get_Op_ptr_from_Vertex(v);
+
+      unit_vector_t qubit_vec;
+
+      bool box_placed = true;
+
+      for (const Command& com : mapping_frontier->circuit_) {
```
[the remark this answers is not in the document]
Comment (sjdilkes) [reply tket/src/Mapping/AASRoute.cpp:74]: alternatively we should be able to trust that the PhasePolyBox is valid from check_method, so cna maybe just delete this.

File: tket/src/Mapping/AASRoute.cpp [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+          }
+        }
+      }
+
+      // check that the box we are working on is really placed and the check
+      // method has been executed
+      // this is imporant if the circuit contains more than one ppb and only one
+      // of them is placed
+
+      if (box_placed) {
```
Comment (sjdilkes) [inline tket/src/Mapping/AASRoute.cpp:90]: Remove if else, trust check_method

File: tket/tests/test_AASRoute.cpp [hunk 3]
```diff
@@ -0,0 +1,429 @@
...
+    circ.rename_units(rename_map);
+
+    std::shared_ptr<MappingFrontier> mf =
+        std::make_shared<MappingFrontier>(circ);
+
+    AASRouteRoutingMethod aasrm(aas::CNotSynthType::Rec, 1);
+
+    REQUIRE(!aasrm.check_method(mf, shared_arc));
+  }
+  GIVEN("test check_method in AASRouteRoutingMethod II") {
```
Comment (sjdilkes) [inline tket/tests/test_AASRoute.cpp:69]: is there a fundamental difference between I and II? May be worth making the titles more descriptive

File: tket/tests/test_AASRoute.cpp [hunk 4]
```diff
@@ -0,0 +1,429 @@
...
+    circ.add_box(ppbox, qubits);
+
+    std::shared_ptr<MappingFrontier> mf =
+        std::make_shared<MappingFrontier>(circ);
+
+    AASRouteRoutingMethod aasrm(aas::CNotSynthType::Rec, 1);
+
+    REQUIRE(!aasrm.check_method(mf, shared_arc));
+  }
+  GIVEN("test routing_method in AASRouteRoutingMethod I") {
```
Comment (sjdilkes) [inline tket/tests/test_AASRoute.cpp:127]: Similarly to other comment, please can you remove I, II and III etc and replace them with something more descriptive?

File: tket/tests/test_AASRoute.cpp [hunk 5]
```diff
@@ -0,0 +1,429 @@
...
+    };
+
+    // this will not use the aas route, which is not clear to me
+    mm.route_circuit(circ, vrm);
+
+    PredicatePtr routed_correctly =
+        std::make_shared<ConnectivityPredicate>(architecture_mixed);
+    PredicatePtrMap preds{CompilationUnit::make_type_pair(routed_correctly)};
+    CompilationUnit cu(circ, preds);
+    REQUIRE(!cu.check_all_predicates());  // TODO
```
Comment (sjdilkes) [inline tket/tests/test_AASRoute.cpp:372]: Todo?

File: tket/tests/test_AASRoute.cpp [hunk 6]
```diff
@@ -0,0 +1,429 @@
...
+
+    PhasePolyBox ppbox(ppb_circ);
+    circ.add_box(ppbox, qubits);
+    circ.add_op<UnitID>(OpType::CX, {qubits[0], qubits[1]});
+
+    circ.rename_units(rename_map);
+
+    Circuit circ_copy(circ);
+
+    std::cout << "\n\nprint out circuit:\n";
```
Comment (sjdilkes) [inline tket/tests/test_AASRoute.cpp:409]: Remove prints

File: tket/tests/test_AASRoute.cpp [hunk 7]
```diff
@@ -0,0 +1,429 @@
...
+
+    for (auto g : circ_copy) {
+      std::cout << g << std::endl;
+    }
+
+    std::cout << "print end circuit\n\n";
+
+    MappingManager mm(shared_arc_mixed);
+
+    // AASRouteRoutingMethod aasrm(aas::CNotSynthType::Rec, 1);
```
Comment (sjdilkes) [inline tket/tests/test_AASRoute.cpp:419]: remove commented out

**Comment by cqc-melf** [comment]
Closed because of merge conflict, reopened as https://github.com/CQCL/tket/pull/254

**Pull request closed without merging**